internal/charmstore/resources.go

```diff
@@ -0,0 +1,255 @@
+	if entity.CharmMeta == nil {
+		return nil, errgo.Newf("entity missing charm metadata")
+	}
+
+	var docs []*mongodoc.Resource
+	for name, meta := range entity.CharmMeta.Resources {
+		doc, err := s.latestResource(entity, channel, name)
+		if errgo.Cause(err) == resourceNotFound {
+			// Stick in a placeholder.
+			doc = &mongodoc.Resource{
+				CharmURL: entity.BaseURL,
+				Name:     meta.Name,
+				// Revision, Fingerprint, etc. are not set.
+			}
+		} else if err != nil {
+			return nil, errgo.Notef(err, "failed to get resource %q", name)
+		}
+		docs = append(docs, doc)
+	}
+	mongodoc.SortResources(docs)
+	return docs, nil
+}
+
+// ResourceBlob holds the information specific to a single resource blob.
+type ResourceBlob struct {
+	io.Reader
+
+	//Fingerprint is the SHA-384 checksum of the blob.
+	Fingerprint []byte
+
+	// Size is the size of the blob in bytes.
+	Size int64
+}
+
+// TODO(ericsnow) We will need Store.nextResourceRevision() to get the
+// value to pass to addResource().
+
+func (s Store) addResource(entity *mongodoc.Entity, name string, blob ResourceBlob, newRevision int) error {
+	if !charmHasResource(entity.CharmMeta, name) {
+		return errgo.Newf("charm does not have resource %q", name)
+	}
+
+	blobName, err := s.storeResource(blob)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(resourceNotFound))
+	}
+
+	doc := &mongodoc.Resource{
+		CharmURL:    entity.BaseURL,
+		Name:        name,
+		Revision:    newRevision,
+		Fingerprint: blob.Fingerprint,
+		Size:        blob.Size,
+		BlobName:    blobName,
+		UploadTime:  time.Now().UTC(),
+	}
+	if s.insertResource(doc); err != nil {
+		if err := s.BlobStore.Remove(blobName); err != nil {
+			logger.Errorf("cannot remove blob %s after error: %v", blobName, err)
+		}
+		return errgo.Mask(err)
+	}
+	return nil
+}
+
+func (s Store) insertResource(doc *mongodoc.Resource) error {
+	if err := doc.Validate(); err != nil {
+		return errgo.Mask(err)
+	}
+	err := s.DB.Resources().Insert(doc)
+	if err != nil && !mgo.IsDup(err) {
+		return errgo.Notef(err, "cannot insert resource")
+	}
+	// TODO(ericsnow) Also fail for dupe?
+	return nil
+}
+
+func (s Store) storeResource(blob ResourceBlob) (string, error) {
+	name := bson.NewObjectId().Hex()
+	// TODO(ericsnow) We will finish this in a follow-up patch.
+	return name, nil
+}
+
+func (s Store) setResource(entity *mongodoc.Entity, channel params.Channel, resName string, revision int) error {
+	if channel == params.NoChannel {
+		return errgo.New("missing channel")
+	}
+	if channel == params.UnpublishedChannel {
+		return errgo.New("cannot publish to unpublished channel")
+	}
+
+	doc, err := s.resource(entity.BaseURL, resName, revision)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(resourceNotFound))
+	}
+
+	if !charmHasResource(entity.CharmMeta, doc.Name) {
+		return errgo.Mask(err)
+	}
+
+	resourcesDoc, err := s.publishedResources(entity.URL, channel)
+	if errgo.Cause(err) == resourceNotFound {
+		resourcesDoc = &mongodoc.Resources{
```

Comment:
add a comment here?  I was confused as to what this code was doing.  Just like

`// no existing published resources, make a new value`